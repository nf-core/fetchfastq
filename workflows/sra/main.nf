/*
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    IMPORT LOCAL MODULES/SUBWORKFLOWS
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
*/

include { MULTIQC_MAPPINGS_CONFIG } from '../../modules/local/multiqc_mappings_config'
include { SRA_FASTQ_FTP           } from '../../modules/local/sra_fastq_ftp'
include { SRA_IDS_TO_RUNINFO      } from '../../modules/local/sra_ids_to_runinfo'
include { SRA_RUNINFO_TO_FTP      } from '../../modules/local/sra_runinfo_to_ftp'
include { ASPERA_CLI              } from '../../modules/local/aspera_cli'
include { SRA_TO_SAMPLESHEET      } from '../../modules/local/sra_to_samplesheet'
include { softwareVersionsToYAML  } from '../../subworkflows/nf-core/utils_nfcore_pipeline'

/*
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    IMPORT NF-CORE SUBWORKFLOWS
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
*/

include { FASTQ_DOWNLOAD_PREFETCH_FASTERQDUMP_SRATOOLS } from '../../subworkflows/nf-core/fastq_download_prefetch_fasterqdump_sratools'

/*
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    IMPORT RECORD TYPES
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
*/

include { Sample } from '../../types/types'

/*
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    RUN MAIN WORKFLOW
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
*/

workflow SRA {

    take:
    ids                         // List<String>
    ena_metadata_fields         // String
    sample_mapping_fields       // String
    nf_core_pipeline            // String
    nf_core_rnaseq_strandedness // String
    download_method             // String enum: 'aspera' | 'ftp' | 'sratools'
    skip_fastq_download         // boolean
    dbgap_key                   // String
    aspera_cli_args             // String
    sra_fastq_ftp_args          // String
    sratools_fasterqdump_args   // String
    sratools_pigz_args          // String
    outdir                      // String

    main:
    ids                                                         // Channel<String>
        //
        // MODULE: Get SRA run information for public database ids
        //
        |> map { id ->
            SRA_IDS_TO_RUNINFO ( id, ena_metadata_fields )
        }                                                       // Channel<Path>
        //
        // MODULE: Parse SRA run information, create file containing FTP links and read into workflow as [ meta, [reads] ]
        //
        |> map(SRA_RUNINFO_TO_FTP)                              // Channel<Path>
        |> set { runinfo_ftp }                                  // Channel<Path>
        |> flatMap { tsv ->
            splitCsv(tsv, header:true, sep:'\t')
        }                                                       // Channel<Map>
        |> map { meta ->
            meta + [single_end: meta.single_end.toBoolean()]
        }                                                       // Channel<Map>
        |> unique                                               // Channel<Map>
        |> set { sra_metadata }                                 // Channel<Map>

    if (!skip_fastq_download) {

        //
        // MODULE: If FTP link is provided in run information then download FastQ directly via FTP and validate with md5sums
        //
        sra_metadata
            |> filter { meta ->
                getDownloadMethod(meta, download_method) == 'ftp'
            }                                                   // Channel<Map>
            |> map { meta ->
                def sample = new Sample( meta, [ file(meta.fastq_1), file(meta.fastq_2) ] )
                SRA_FASTQ_FTP ( sample, sra_fastq_ftp_args )
            }                                                   // fastq: Channel<Sample>, md5: Channel<Sample>
            |> set { ftp_samples }                              // fastq: Channel<Sample>, md5: Channel<Sample>

        //
        // SUBWORKFLOW: Download sequencing reads without FTP links using sra-tools.
        //
        sra_metadata
            |> filter { meta ->
                getDownloadMethod(meta, download_method) == 'sratools'
            }                                                   // Channel<Map>
            |> map { meta ->
                new Tuple2<Map,String>( meta, meta.run_accession )
            }                                                   // Channel<Tuple2<Map,String>>
            |> FASTQ_DOWNLOAD_PREFETCH_FASTERQDUMP_SRATOOLS (
                dbgap_key ? file(dbgap_key, checkIfExists: true) : [],
                sratools_fasterqdump_args,
                sratools_pigz_args )                            // Channel<Sample>
            |> set { sratools_samples }                         // Channel<Sample>
 
        //
        // MODULE: If Aspera link is provided in run information then download FastQ directly via Aspera CLI and validate with md5sums
        //
        sra_metadata
            |> filter { meta ->
                getDownloadMethod(meta, download_method) == 'aspera'
            }                                                   // Channel<Map>
            |> map { meta ->
                def sample = new Sample( meta, meta.fastq_aspera.tokenize(';').take(2).collect( name -> file(name) ) )
                ASPERA_CLI ( sample, 'era-fasp', aspera_cli_args )
            }                                                   // fastq: Channel<Sample>, md5: Channel<Sample>
            |> set { aspera_samples }                           // fastq: Channel<Sample>, md5: Channel<Sample>

        // Isolate FASTQ channel which will be added to emit block
        fastq = mix(ftp_samples.fastq, sratools_samples.reads, aspera_samples.fastq)
        md5 = mix(ftp_samples.md5, aspera_samples.md5)

        fastq                                                   // Channel<Sample>
            |> map { sample ->
                def reads = sample.files
                def meta = sample.meta
                meta + [
                    fastq_1: reads[0] ? "${outdir}/fastq/${reads[0].getName()}" : '',
                    fastq_2: reads[1] && !meta.single_end ? "${outdir}/fastq/${reads[1].getName()}" : ''
                ]
            }                                                   // Channel<Map>
            |> set { sra_metadata }                             // Channel<Map>
    }

    //
    // MODULE: Stage FastQ files downloaded by SRA together and auto-create a samplesheet
    //
    sra_metadata                                            // Channel<Map>
        |> collect                                          // List<Map>
        |> { sra_metadata ->
            SRA_TO_SAMPLESHEET (
                sra_metadata,
                nf_core_pipeline,
                nf_core_rnaseq_strandedness,
                sample_mapping_fields )
        }                                                   // samplesheet: Path, mappings: Path
        |> set { index_files }                              // samplesheet: Path, mappings: Path

    samplesheet = index_files.samplesheet                   // Path
    mappings    = index_files.mappings                      // Path

    //
    // MODULE: Create a MutiQC config file with sample name mappings
    //
    sample_mappings = sample_mapping_fields
        ? MULTIQC_MAPPINGS_CONFIG ( mappings )              // Path
        : null

    //
    // Collate and save software versions
    //
    'versions'                                              // String
        |> Channel.topic                                    // Channel<Tuple3<String,String,String>>
        |> softwareVersionsToYAML                           // Channel<String>
        |> collect(sort: true)                              // List<String>
        |> exec('SOFTWARE_VERIONS') { versions ->
            def path = task.workDir.resolve('nf_core_fetchngs_software_mqc_versions.yml')
            mergeText(versions, path, newLine: true)
            return path
        }                                                   // Path
        |> set { versions_yml }                             // Path

    emit:
    samplesheet
    mappings
    sample_mappings
    sra_metadata

    publish:
    fastq           >> 'fastq/'
    md5             >> 'fastq/md5/'
    runinfo_ftp     >> 'metadata/'
    versions_yml    >> 'pipeline_info/'
    samplesheet     >> 'samplesheet/'
    mappings        >> 'samplesheet/'
    sample_mappings >> 'samplesheet/'
}

/*
========================================================================================
    FUNCTIONS
========================================================================================
*/

def getDownloadMethod(Map meta, String download_method) {
    // meta.fastq_aspera is a metadata string with ENA fasp links supported by Aspera
        // For single-end: 'fasp.sra.ebi.ac.uk:/vol1/fastq/ERR116/006/ERR1160846/ERR1160846.fastq.gz'
        // For paired-end: 'fasp.sra.ebi.ac.uk:/vol1/fastq/SRR130/020/SRR13055520/SRR13055520_1.fastq.gz;fasp.sra.ebi.ac.uk:/vol1/fastq/SRR130/020/SRR13055520/SRR13055520_2.fastq.gz'
    if (meta.fastq_aspera && download_method == 'aspera')
        return 'aspera'
    if ((!meta.fastq_aspera && !meta.fastq_1) || download_method == 'sratools')
        return 'sratools'
    return 'ftp'
}

/*
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    THE END
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
*/
