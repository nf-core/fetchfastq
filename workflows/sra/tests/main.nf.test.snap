{
    "Parameters: default": {
        "content": [
            "samplesheet.csv",
            [
                "id_mappings.csv:md5,3e41ce6ab19feb76f2b20fa77a910ad3"
            ],
            [
                "multiqc_config.yml:md5,1ac06bb95b503703430e74660bbdd768"
            ],
            [
                {
                    "base_count": "194930",
                    "experiment_accession": "DRX026011",
                    "experiment_alias": "DRX026011",
                    "experiment_title": "Illumina HiSeq 2500 paired end sequencing: Illumina HiSeq 2500 paired end sequencing of SAMD00024405",
                    "fastq_1": "DRX026011_DRR028935_1.fastq.gz:md5,f9b4a5d25aa7d6ab4050dbe0764eb04d",
                    "fastq_2": "DRX026011_DRR028935_2.fastq.gz:md5,dc0d2e1fc46f4d4581a477f87ec3ff15",
                    "fastq_aspera": "fasp.sra.ebi.ac.uk:/vol1/fastq/DRR028/DRR028935/DRR028935_1.fastq.gz;fasp.sra.ebi.ac.uk:/vol1/fastq/DRR028/DRR028935/DRR028935_2.fastq.gz",
                    "fastq_bytes": "60275;61610",
                    "fastq_ftp": "ftp.sra.ebi.ac.uk/vol1/fastq/DRR028/DRR028935/DRR028935_1.fastq.gz;ftp.sra.ebi.ac.uk/vol1/fastq/DRR028/DRR028935/DRR028935_2.fastq.gz",
                    "fastq_galaxy": "ftp.sra.ebi.ac.uk/vol1/fastq/DRR028/DRR028935/DRR028935_1.fastq.gz;ftp.sra.ebi.ac.uk/vol1/fastq/DRR028/DRR028935/DRR028935_2.fastq.gz",
                    "fastq_md5": "cc01df82a9354bb6b3be93483b20c35a;941c9998d746416dad53c94c480ddf30",
                    "id": "DRX026011_DRR028935",
                    "instrument_model": "Illumina HiSeq 2500",
                    "instrument_platform": "ILLUMINA",
                    "library_layout": "PAIRED",
                    "library_name": "day0_BbSQE-I",
                    "library_selection": "cDNA",
                    "library_source": "TRANSCRIPTOMIC",
                    "library_strategy": "RNA-Seq",
                    "md5_1": "cc01df82a9354bb6b3be93483b20c35a",
                    "md5_2": "941c9998d746416dad53c94c480ddf30",
                    "read_count": "965",
                    "run_accession": "DRR028935",
                    "run_alias": "DRR028935",
                    "sample_accession": "SAMD00024405",
                    "sample_alias": "SAMD00024405",
                    "sample_description": "Liquid culture",
                    "sample_title": "Botryococcus braunii Showa at day 0 after inoculation into fresh culture medium",
                    "scientific_name": "Botryococcus braunii Showa",
                    "secondary_sample_accession": "DRS019431",
                    "secondary_study_accession": "DRP002616",
                    "single_end": false,
                    "study_accession": "PRJDB3420",
                    "study_alias": "DRP002616",
                    "study_title": "Liquid culture of Botryococcus braunii, race B, Showa",
                    "submission_accession": "DRA002949",
                    "tax_id": "1202541"
                },
                {
                    "base_count": "1996273",
                    "experiment_accession": "SRX6725035",
                    "experiment_alias": "Emb289P1_bin131",
                    "experiment_title": "Illumina HiSeq 2500 sequencing: Binning of metagenomic reads from the P1 gut compartment of Embiratermes neotenicus",
                    "fastq_1": "SRX6725035_SRR9984183.fastq.gz:md5,c9b4c2c4abf19da7082caf442fdefea7",
                    "fastq_2": "",
                    "fastq_aspera": "fasp.sra.ebi.ac.uk:/vol1/fastq/SRR998/003/SRR9984183/SRR9984183.fastq.gz",
                    "fastq_bytes": "605358",
                    "fastq_ftp": "ftp.sra.ebi.ac.uk/vol1/fastq/SRR998/003/SRR9984183/SRR9984183.fastq.gz",
                    "fastq_galaxy": "ftp.sra.ebi.ac.uk/vol1/fastq/SRR998/003/SRR9984183/SRR9984183.fastq.gz",
                    "fastq_md5": "0b512d2dc31685983456bd56fd836544",
                    "id": "SRX6725035_SRR9984183",
                    "instrument_model": "Illumina HiSeq 2500",
                    "instrument_platform": "ILLUMINA",
                    "library_layout": "SINGLE",
                    "library_name": "Emb289P1_bin131",
                    "library_selection": "RANDOM",
                    "library_source": "METAGENOMIC",
                    "library_strategy": "WGS",
                    "md5_1": "0b512d2dc31685983456bd56fd836544",
                    "md5_2": "",
                    "read_count": "58",
                    "run_accession": "SRR9984183",
                    "run_alias": "Emb289P1_bin131.fastq",
                    "sample_accession": "SAMN12581720",
                    "sample_alias": "Emb289P1_bin131",
                    "sample_description": "Keywords: GSC:MIxS MIMAG:5.0",
                    "sample_title": "MIMAG Metagenome-assembled Genome sample from Defluviitaleaceae bacterium",
                    "scientific_name": "Defluviitaleaceae bacterium",
                    "secondary_sample_accession": "SRS5277011",
                    "secondary_study_accession": "SRP218535",
                    "single_end": true,
                    "study_accession": "PRJNA560329",
                    "study_alias": "PRJNA560329",
                    "study_title": "Phylogenomic analysis of 589 metagenome-assembled genomes encompassing all major prokaryotic lineages from the gut of higher termites",
                    "submission_accession": "SRA942061",
                    "tax_id": "2660712"
                },
                {
                    "base_count": "35658",
                    "experiment_accession": "ERX1234253",
                    "experiment_alias": "qiita_ptid_1263:10317.BLANK.93.3E.r22",
                    "experiment_title": "Illumina HiSeq 2500 sequencing: qiita_ptid_1263:10317.BLANK.93.3E.r22",
                    "fastq_1": "ERX1234253_ERR1160846.fastq.gz:md5,b1de59f829fbd2d60a1ed4bc3da55709",
                    "fastq_2": "",
                    "fastq_aspera": "fasp.sra.ebi.ac.uk:/vol1/fastq/ERR116/006/ERR1160846/ERR1160846.fastq.gz",
                    "fastq_bytes": "18077",
                    "fastq_ftp": "ftp.sra.ebi.ac.uk/vol1/fastq/ERR116/006/ERR1160846/ERR1160846.fastq.gz",
                    "fastq_galaxy": "ftp.sra.ebi.ac.uk/vol1/fastq/ERR116/006/ERR1160846/ERR1160846.fastq.gz",
                    "fastq_md5": "5924f20ef547ebdfed7cad795bbab6e6",
                    "id": "ERX1234253_ERR1160846",
                    "instrument_model": "Illumina HiSeq 2500",
                    "instrument_platform": "ILLUMINA",
                    "library_layout": "SINGLE",
                    "library_name": "10317.BLANK.93.3E.r22",
                    "library_selection": "PCR",
                    "library_source": "METAGENOMIC",
                    "library_strategy": "AMPLICON",
                    "md5_1": "5924f20ef547ebdfed7cad795bbab6e6",
                    "md5_2": "",
                    "read_count": "283",
                    "run_accession": "ERR1160846",
                    "run_alias": "qiita_ppdid_706:10317.BLANK.93.3E.r22",
                    "sample_accession": "SAMEA3687214",
                    "sample_alias": "qiita_sid_10317:10317.BLANK.93.3E.r22",
                    "sample_description": "American Gut control",
                    "sample_title": "10317.BLANK.93.3E.r22",
                    "scientific_name": "metagenome",
                    "secondary_sample_accession": "ERS994363",
                    "secondary_study_accession": "ERP012803",
                    "single_end": true,
                    "study_accession": "PRJEB11419",
                    "study_alias": "qiita_sid_10317",
                    "study_title": "American Gut Project",
                    "submission_accession": "ERA541392",
                    "tax_id": "256318"
                }
            ],
            [
                "versions.yml:md5,1496d1cbc9041e07ab8a0c25f0b054d9",
                "versions.yml:md5,9b17045ca8bdc272cb3f9d349a81d206",
                "versions.yml:md5,b52279f7d6b891a6523d9321f3f85b47",
                "versions.yml:md5,c19fc0f920e57875881761cc944a46d7"
            ]
        ],
        "meta": {
            "nf-test": "0.9.0",
            "nextflow": "24.09.2"
        },
        "timestamp": "2024-10-15T13:35:40.78967"
    }
}